### ISSUE TITLE
Enforce --offline switch in run_tests.py

### ISSUE THREAD

**Issue by peterjc** [biopython/biopython#846]
Can we do something like monkey-patching urllib for `run_tests.py --offline` so that attempting use an online resource triggers our own exception making the tests fail? This would have caught #819 much earlier for example, and might help if for example Linux packagers try to run the test suite in a network-less chroot jail or similar.

**Comment by Gasta88** [comment]
Hi,
could you elaborate a bit better the issue ? What do you mean with "monkey-patching"?

**Comment by peterjc** [comment]
Changing a function at run time is often called "monkey patching" in Python, https://stackoverflow.com/questions/5626193/what-is-monkey-patching

We could (within ``run_tests.py``) do something like this to replace the ``urlopen`` function (etc) with a stub which just raise an exception:

```python
from Bio._py3k import urlopen

def dummy_urlopen(...):
    raise RuntimeError("Bad test! You are not allowed to go online!")

urlopen = dummy_urlopen  # replace with error raising stub!
```

I've not tested this - but something like that ought to work.

**Comment by Gasta88** [comment]
Thanks, I wasn't sure that "monkey patching" was a real thing :)

I'll take a look at this.

**Comment by peterjc** [comment]
My idea was if ``run_tests.py`` is called with ``--offline``, then it should do the monkey patching to disable ``urlopen`` (etc).

Any test known to need the network should be using this code snippet, which means it get skipped if ``--offline`` is used:

```python
import requires_internet
requires_internet.check()
```

Any test using the network anyway would then fail (because of the altered ``urlopen``), so we'd know about this problem, and add that snippet of code marking the test as needing to go online.

**Comment by peterjc** [comment]
Closed via #1643, thank you.
